docs/proposals/20210524-managed-external-etcd.md

```diff
@@ -0,0 +1,570 @@
+- An etcd member will need to be replaced by the etcd controller in two cases:
+  - Member is found unhealthy
+  - During an upgrade
+- While replacing an unhealthy etcd member, it is important to first remove the member and only then replace it by adding a new one. This ensures the cluster stays stable. [This Etcd doc](https://etcd.io/docs/v3.4/faq/#should-i-add-a-member-before-removing-an-unhealthy-member) explains why removing an unhealthy member is important in depth. Etcd's latest versions allow adding new members as learners so it doesn't affect the quorum size, but that feature is in beta. Also, adding a learner is a two-step process, where you first add a member as learner and then promote it, and etcdadm currently doesn't support it. So we can include this in a later version once the learner feature becomes GA and etcdadm incorporates the required changes.
+- We can reuse this same "scale down" function when removing a healthy member during an etcd upgrade. So even upgrade will follow the same pattern of removing the member first and then adding a new one with the latest spec.
+- Once the upgrade/machine replacement is completed, the etcd controller will again update the EtcdadmCluster.Status.Endpoints field. The control plane controller will rollout new Machines that use the updated etcd endpoints for the apiserver.
+
+###### Periodic etcd member healthchecks
+- The etcdadm cluster controller will perform healthchecks on the etcd members periodically at a predetermined interval. The controller will perform client healthchecks by making HTTP Get requests to the `<etcd member address>:2379/health` endpoint of each etcd member. It cannot perform peer-to-peer healthchecks since the controller is external to the etcd cluster and there won't be any agents running on the etcd members.
+- If a member continuously fails healthcheck for a certain amount of time, the etcdadm cluster controller will replace that member with a new member.
+
+#### Changes needed in docker machine controller
+- Each machine infrastructure provider sets a providerID on every InfraMachine.
+- Infrastructure providers like CAPA and CAPV set this provider ID based on the instance metadata.
+- CAPD on the other hand first calls `kubectl patch` to add the provider ID on the node. And only then it sets the provider ID value on the DockerMachine resource. But the problem is that in case of an etcd only cluster, the machine is not registered as a Kubernetes node. Cluster provisioning does not progress until this provider ID is set on the Machine. As a solution, CAPD will check if the bootstrap provider is etcdadm, and skip the kubectl patch process and set the provider ID on DockerMachine directly. These changes are required in the [docker machine controller](https://github.com/mrajashree/cluster-api/pull/2/files#diff-1923dd8291a9406e3c144763f526bd9797a2449a030f5768178b8d06d13c795bR307) for CAPD.
```

Comment:
I'm -1 on this change.
Each provider should be sandboxed and rely only on informations available on CAPI resources (vs checking combinations with other providers). If with this proposal we are introducing the concept of machine without a node, let's find a way to make this is first class construct in CAPI core. This can eventually be done in staged approaches avoiding breaking changes in the beginning, but at the end we should provide the users a way to discriminate the different types of machines.